modules/animation/src/animationcontroller.cpp

```diff
@@ -469,46 +303,88 @@ void AnimationController::tick() {
     setState(newState);
 }
 
-void AnimationController::tickRender() {
-    if (state_ != AnimationState::Rendering) {
-        setState(AnimationState::Paused);
-        return;
-    }
+void AnimationController::render() {
+    auto start = std::chrono::high_resolution_clock::now();
 
-    // Save all active canvases from last rendering tick:
-    // Apparently, we need to be outside of tickRender() to let the system refresh its state
-    // The first call to tickRender() is done with renderState_.currentFrame == -1 to bring the
-    // system to a proper state
-    // - generate filename pattern
-    if (renderState_.currentFrame >= 0) {
-        std::stringstream fileNamePattern;
-        fileNamePattern << renderBaseName.get() << renderState_.canvasIndicator << std::setfill('0')
-                        << std::setw(renderState_.digits) << renderState_.currentFrame;
-        auto ext = FileExtension::createFileExtensionFromString(renderImageExtension.get());
-        // - save active canvases
-        util::saveAllCanvases(app_->getProcessorNetwork(), renderLocation.get(),
-                              fileNamePattern.str(), ext.extension_, true);
-    }
+    auto network = app_->getProcessorNetwork();
+    std::optional<NetworkLock> lock{network};
 
-    // Next!
-    renderState_.currentFrame++;
+    setState(AnimationState::Rendering);
 
-    // Are we finished?
-    if (renderState_.currentFrame >= renderState_.numFrames) {
+    // Gather rendering info
+    const Seconds firstTime =
+        (renderWindowMode.get() == 0) ? animation_->getFirstTime() : Seconds(renderWindow.get()[0]);
+    const Seconds lastTime =
+        (renderWindowMode.get() == 0) ? animation_->getLastTime() : Seconds(renderWindow.get()[1]);
+    const int numFrames =
+        std::max(2, static_cast<int>((lastTime - firstTime) / Seconds{1.0 / renderFPS.get()}));
+
+    // digits of the frame counter
+    const int digits = [&]() {
+        int d = 0;
+        int number(numFrames - 1);
+        while (number) {
+            number /= 10;
+            d++;
+        }
+        // use at least 4 digits, so we nicely overwrite the files from a previous test rendering
+        // with less frames
+        return std::max(d, 4);
+    }();
+
+    // Get all active canvases
+    std::vector<std::pair<Exporter*, std::string>> exporters;
+    network->forEachProcessor([&](Processor* p) {
+        if (p->isSink()) {
+            if (auto exporter = dynamic_cast<Exporter*>(p)) {
+                auto base = renderBaseName.get();
+                replaceInString(base, "UPN", p->getIdentifier());
```

Comment:
Where does this UPN come from and what is its purpose?

Reply:
Think it's an old invention by Rickard. it means "use processor name"
There are probably better ways to do that... but that's how we have done it for a long time...